File: app/web/src/pages/chat/outboxStore.ts
// Persisted send outbox (protocol v2). One entry per in-flight send, keyed
// by `platform_msg_id`, persisted to localStorage under
// `baybo.outbox.<sessionId>` (same pattern as `baybo.queue.<sessionId>`) so
// a tab close / reload keeps un-durable sends alive.
//
// Confirmation is two-stage (docs/sync-protocol.md "Send path"):
//   * the server's Echo (ordinal-less user Message with the same
//     platform_msg_id) proves TRANSPORT — `sending` → `sent`, in-connection
//     retries stop;
//   * an ordinal-stamped row with the same platform_msg_id (sync / backfill)
//     proves DURABILITY and releases the entry.
// The retry machine: no echo within ECHO_TIMEOUT_MS → one blind resend
// (same key; safe inside the gateway's live dedup window), hard-capped at
// MAX_AUTO_TRANSMISSIONS total, then `failed` + manual retry. A rebased sync
// hides the floor, so unconfirmed entries go `unknown` and resolve via the
// per-key point lookup instead of blind-resending.

import type { WireAttachment } from '../../api/chatWs';

export type OutboxEntryState = 'sending' | 'sent' | 'failed' | 'unknown';

export interface OutboxEntry {
  platformMsgId: string;
  text: string;
  attachments?: WireAttachment[];
  state: OutboxEntryState;
  /** Automatic transmissions so far (initial send counts as 1). */
  transmissions: number;
  createdAt: number;
  lastSentAt: number;
}

export const OUTBOX_KEY_PREFIX = 'baybo.outbox.';
/** No Echo within this window (state still `sending`, connection alive)
 *  triggers the one blind in-connection resend. Durability lag alone never
 *  does — a mid-turn persist legitimately takes minutes. */
export const ECHO_TIMEOUT_MS = 10_000;
/** Hard cap on automatic transmissions per message (initial + retries);
 *  past it the entry goes `failed` and only a manual retry resets it. */
export const MAX_AUTO_TRANSMISSIONS = 3;

const outboxKey = (sessionId: string) => `${OUTBOX_KEY_PREFIX}${sessionId}`;

/** Whether a `sending` entry's echo window has lapsed and it is still
 *  under the automatic-transmission cap — i.e. a blind resend is due. */
export function dueForBlindResend(entry: OutboxEntry, now: number): boolean {
  return (
    entry.state === 'sending' &&
    entry.transmissions < MAX_AUTO_TRANSMISSIONS &&
    now - entry.lastSentAt >= ECHO_TIMEOUT_MS
  );
}

/** Whether a `sending` entry has exhausted its cap without an echo —
 *  i.e. it must flip to `failed`. */
export function resendExhausted(entry: OutboxEntry, now: number): boolean {
  return (
    entry.state === 'sending' &&
    entry.transmissions >= MAX_AUTO_TRANSMISSIONS &&
    now - entry.lastSentAt >= ECHO_TIMEOUT_MS
  );
}

const OUTBOX_STATES: readonly OutboxEntryState[] = ['sending', 'sent', 'failed', 'unknown'];

function isOutboxEntry(v: unknown): v is OutboxEntry {
  if (typeof v !== 'object' || v === null) return false;
  const e = v as Record<string, unknown>;
  return (
    typeof e.platformMsgId === 'string' &&
    e.platformMsgId.length > 0 &&
    typeof e.text === 'string' &&
    (e.attachments === undefined || Array.isArray(e.attachments)) &&
    OUTBOX_STATES.includes(e.state as OutboxEntryState) &&
    typeof e.transmissions === 'number' &&
    typeof e.createdAt === 'number' &&
    typeof e.lastSentAt === 'number'
  );
}

function readEntries(sessionId: string): Map<string, OutboxEntry> {
  const out = new Map<string, OutboxEntry>();
  try {
    const raw = window.localStorage.getItem(outboxKey(sessionId));
    if (raw === null) return out;
    const parsed: unknown = JSON.parse(raw);
    if (!Array.isArray(parsed)) return out;
    for (const item of parsed) {
      if (isOutboxEntry(item)) out.set(item.platformMsgId, item);
    }
  } catch {
    /* storage disabled / corrupt — in-memory tracking still works */
  }
  return out;
}

/** Imperative per-tab outbox handle. Not reactive — the transcript row's
 *  pending/failed flags are driven by the caller alongside each mutation. */
export class OutboxStore {
  private cache = new Map<string, Map<string, OutboxEntry>>();

  private session(sessionId: string): Map<string, OutboxEntry> {
    let entries = this.cache.get(sessionId);
    if (!entries) {
      entries = readEntries(sessionId);
      this.cache.set(sessionId, entries);
    }
    return entries;
  }

  private persist(sessionId: string): void {
    const entries = this.session(sessionId);
    try {
      if (entries.size === 0) {
        window.localStorage.removeItem(outboxKey(sessionId));
      } else {
        window.localStorage.setItem(outboxKey(sessionId), JSON.stringify([...entries.values()]));
      }
    } catch {
      /* storage disabled */
    }
  }

  private update(
    sessionId: string,
    platformMsgId: string,
    fn: (entry: OutboxEntry) => OutboxEntry,
  ): OutboxEntry | undefined {
    const entries = this.session(sessionId);
    const cur = entries.get(platformMsgId);
    if (!cur) return undefined;
    const next = fn(cur);
    entries.set(platformMsgId, next);
    this.persist(sessionId);
    return next;
  }

  entries(sessionId: string): OutboxEntry[] {
    return [...this.session(sessionId).values()];
  }

  get(sessionId: string, platformMsgId: string): OutboxEntry | undefined {
    return this.session(sessionId).get(platformMsgId);
  }

  /** Every session that has at least one live entry — the union of ids
   *  already cached this page-load and ids persisted by a prior one. */
  sessionIds(): string[] {
    const ids = new Set<string>();
    for (const [sid, entries] of this.cache) {
      if (entries.size > 0) ids.add(sid);
    }
    try {
      for (let i = 0; i < window.localStorage.length; i++) {
        const key = window.localStorage.key(i);
        if (key?.startsWith(OUTBOX_KEY_PREFIX)) ids.add(key.slice(OUTBOX_KEY_PREFIX.length));
      }
    } catch {
      /* storage disabled */
    }
    return [...ids];
  }

  /** Every platform_msg_id not yet durability-confirmed (any state — a
   *  `failed` row must stay on screen with its retry affordance). Drives
   *  the REPLACE-overlay rule. */
  unconfirmedIds(sessionId: string): Set<string> {
    return new Set(this.session(sessionId).keys());
  }

  /** Record a fresh send: state `sending`, the transmission already counted
   *  (callers only enroll sends that actually hit the wire). */
  beginSend(
    sessionId: string,
    input: { platformMsgId: string; text: string; attachments?: WireAttachment[] },
    now = Date.now(),
  ): OutboxEntry {
    const entry: OutboxEntry = {
      platformMsgId: input.platformMsgId,
      text: input.text,
      attachments: input.attachments && input.attachments.length > 0 ? input.attachments : undefined,
      state: 'sending',
      transmissions: 1,
      createdAt: now,
      lastSentAt: now,
    };
    this.session(sessionId).set(entry.platformMsgId, entry);
    this.persist(sessionId);
    return entry;
  }

  /** Echo observed: transport confirmed, in-connection retries stop. The
   *  entry is retained until durability confirms. */
  markEchoed(sessionId: string, platformMsgId: string): OutboxEntry | undefined {
    return this.update(sessionId, platformMsgId, (e) =>
      e.state === 'sending' || e.state === 'unknown' ? { ...e, state: 'sent' } : e,
    );
  }

  /** An ordinal-stamped row with this key was observed (sync / backfill /
   *  point lookup): durability confirmed, release the entry. Returns whether
   *  an entry existed (so callers can clear a row's pending/failed flags). */
  confirmDurable(sessionId: string, platformMsgId: string): boolean {
    const entries = this.session(sessionId);
    if (!entries.delete(platformMsgId)) return false;
    this.persist(sessionId);
    return true;
  }

  /** Count one (re)transmission: state back to `sending`, echo timer re-armed. */
  recordTransmission(sessionId: string, platformMsgId: string, now = Date.now()): OutboxEntry | undefined {
    return this.update(sessionId, platformMsgId, (e) => ({
      ...e,
      state: 'sending',
      transmissions: e.transmissions + 1,
      lastSentAt: now,
    }));
  }

  markFailed(sessionId: string, platformMsgId: string): OutboxEntry | undefined {
    return this.update(sessionId, platformMsgId, (e) => ({ ...e, state: 'failed' }));
  }

  /** A rebased page hid the floor — this entry's absence is unknowable from
   *  the page alone. Park it `unknown` pending the point lookup. */
  markUnknown(sessionId: string, platformMsgId: string): OutboxEntry | undefined {
    return this.update(sessionId, platformMsgId, (e) =>
      e.state === 'failed' ? e : { ...e, state: 'unknown' },
    );
  }

  /** Point lookup proved the key absent: the normal retry machine resumes.
   *  `lastSentAt` is zeroed so the next retry tick treats the entry as due
   *  immediately; no transmission is consumed by the lookup itself. */
  resumeSending(sessionId: string, platformMsgId: string): OutboxEntry | undefined {
    return this.update(sessionId, platformMsgId, (e) => ({
      ...e,
      state: 'sending',
      lastSentAt: 0,
    }));
  }

  /** Manual retry from the failed affordance: same platform_msg_id, the
   *  automatic-transmission budget resets. */
  resetForManualRetry(sessionId: string, platformMsgId: string): OutboxEntry | undefined {
    return this.update(sessionId, platformMsgId, (e) => ({
      ...e,
      state: 'sending',
      transmissions: 0,
      lastSentAt: 0,
    }));
  }
}

File: app/ios/App/Core/OutboxStore.swift
import Foundation

/// Persisted send outbox (sync-v2). One entry per in-flight send, keyed by its
/// `platform_msg_id`, persisted to `Application Support/baybo/outbox/<id>.json`
/// (a sibling of the transcript mirror) so a relaunch keeps un-durable sends
/// alive.
///
/// Confirmation is two-stage (docs/sync-protocol.md "Send path"): the server's
/// Echo (ordinal-less user message with the same key) proves TRANSPORT
/// (`sending` → `sent`, in-connection retries stop); an ordinal-stamped row
/// with the same key (from sync / backfill) proves DURABILITY and releases the
/// entry. The retry machine: no echo within 10 s → one blind resend (safe
/// inside the gateway's live dedup window), hard-capped at 3 transmissions
/// total, then `failed` + the manual retry affordance. A rebased sync hides the
/// floor, so an unconfirmed entry goes `unknown` and resolves via the per-key
/// point lookup instead of blind-resending.
///
/// The outbox rides out a BRIEF outage silently (a spinner, not a failure — the
/// message is queued, and reconnect drains it). But a SUSTAINED one is a real
/// failure the user must see: an entry still `sending` past `queuedTimeout`
/// (wall-clock, no wire in reach) flips to `failed` too, so an offline send
/// surfaces the red dot rather than spinning forever. The reconnect gate then
/// leaves it alone — it never silently auto-sends after being declared failed.
enum OutboxEntryState: String, Codable {
    case sending
    case sent
    case failed
    case unknown
}

/// Blob-ref metadata for an outbox entry's attachments — a Codable mirror of
/// the FFI `AttachmentRef` so a retry/resend replays the SAME image/file
/// message, not a text-only one. The blob bytes stay server-side
/// (content-addressed), so the ref is all a resend needs.
struct OutboxAttachment: Codable {
    /// `"image"` / `"audio"` / `"file"` (the `AttachmentKind` discriminant).
    var kind: String
    var blobId: String
    var mimeType: String
    var size: UInt32
    var filename: String?
}

struct OutboxEntry: Codable {
    let platformMsgId: String
    var text: String
    var attachments: [OutboxAttachment]
    var state: OutboxEntryState
    /// Automatic transmissions so far (the initial send counts as 1).
    var transmissions: Int
    /// Epoch seconds of the entry's creation and its most recent transmission.
    var createdAt: Double
    var lastSentAt: Double
}

@MainActor
final class OutboxStore {
    /// No Echo within this window (state still `sending`, connection alive)
    /// triggers the one blind in-connection resend. Durability lag alone never
    /// does — a mid-turn persist legitimately takes minutes.
    static let echoTimeout: TimeInterval = 10
    /// Hard cap on automatic transmissions per message (initial + retries); past
    /// it the entry goes `failed` and only a manual retry resets it.
    static let maxAutoTransmissions = 3
    /// Retry-sweep cadence — fine-grained enough to fire within a few seconds of
    /// the 10 s no-echo deadline without busy-spinning.
    static let retryTick: Duration = .seconds(3)
    /// Wall-clock lifetime cap for a still-unsent (`sending`) entry on a DOWN
    /// leg. Past it a queued message is declared `failed` rather than left
    /// spinning — a sustained outage IS a send failure. Sized at (not under) the
    /// live-leg genuine-failure horizon (≈ `maxAutoTransmissions × echoTimeout`)
    /// so it targets the offline case and never preempts the connected retry.
    static let queuedTimeout: TimeInterval = 30

    private let sessionId: String
    private let directory: URL
    /// The retry machine's clock. Injected so the 10 s echo window is testable
    /// without sleeping through it.
    private let now: () -> Date
    private var items: [String: OutboxEntry]

    convenience init(sessionId: String) {
        self.init(
            sessionId: sessionId,
            directory: Self.directory(in: SessionIndex.supportDirectory()),
            now: Date.init)
    }

    init(sessionId: String, directory: URL, now: @escaping () -> Date) {
        self.sessionId = sessionId
        self.directory = directory
        self.now = now
        items = Self.read(sessionId: sessionId, in: directory)
    }

    // MARK: - Reads

    /// Every live entry (any state — a `failed` row stays with its retry
    /// affordance until released).
    func entries() -> [OutboxEntry] {
        Array(items.values)
    }

    /// Whether a `sending` entry's echo window has lapsed and it is still under
    /// the automatic-transmission cap — i.e. a blind resend is due.
    func dueForBlindResend(_ entry: OutboxEntry) -> Bool {
        entry.state == .sending
            && entry.transmissions < Self.maxAutoTransmissions
            && echoWindowLapsed(entry)
    }

    /// Whether a `sending` entry exhausted its cap without an echo — it must flip
    /// to `failed`.
    func resendExhausted(_ entry: OutboxEntry) -> Bool {
        entry.state == .sending
            && entry.transmissions >= Self.maxAutoTransmissions
            && echoWindowLapsed(entry)
    }

    private func echoWindowLapsed(_ entry: OutboxEntry) -> Bool {
        now().timeIntervalSince1970 - entry.lastSentAt >= Self.echoTimeout
    }

    /// Whether a still-unsent entry has outlived the queued-lifetime cap — a
    /// sustained down leg that should now surface as a failure. Only a `sending`
    /// entry qualifies: an echoed (`sent`) entry already proved transport and is
    /// merely awaiting durability, so a stalled mid-turn persist never fails it.
    func queuedTooLong(_ entry: OutboxEntry) -> Bool {
        entry.state == .sending
            && now().timeIntervalSince1970 - entry.createdAt >= Self.queuedTimeout
    }

    /// Session ids with a persisted outbox. A relaunch reads this to find sends
    /// the previous process enrolled — including for a session that never
    /// reached `SessionIndex`, which nothing else can reach
    /// (`AppStore.resumeStrandedSends`). Session ids are UUIDs, so the file
    /// name's `/` sanitisation never fires and the basename IS the id.
    nonisolated static func pendingSessionIds(in supportDirectory: URL) -> [String] {
        let contents =
            (try? FileManager.default.contentsOfDirectory(
                at: directory(in: supportDirectory), includingPropertiesForKeys: nil)) ?? []
        return
            contents
            .filter { $0.pathExtension == "json" }
            .map { $0.deletingPathExtension().lastPathComponent }
    }

    // MARK: - Mutations

    /// Record a fresh send: state `sending`, the transmission already counted
    /// (only sends that actually hit the wire are enrolled).
    func beginSend(platformMsgId: String, text: String, attachments: [OutboxAttachment]) {
        let at = now().timeIntervalSince1970
        items[platformMsgId] = OutboxEntry(
            platformMsgId: platformMsgId, text: text, attachments: attachments,
            state: .sending, transmissions: 1, createdAt: at, lastSentAt: at)
        persist()
    }

    /// Echo observed: transport confirmed, in-connection retries stop. The entry
    /// is retained until durability confirms.
    func markEchoed(platformMsgId: String) {
        mutate(platformMsgId) { entry in
            if entry.state == .sending || entry.state == .unknown { entry.state = .sent }
        }
    }

    /// An ordinal-stamped row with this key was observed (sync / backfill / point
    /// lookup): durability confirmed, release the entry.
    func confirmDurable(platformMsgId: String) {
        guard items.removeValue(forKey: platformMsgId) != nil else { return }
        persist()
    }

    /// Count one (re)transmission: state back to `sending`, echo timer re-armed.
    func recordTransmission(platformMsgId: String) {
        let at = now().timeIntervalSince1970
        mutate(platformMsgId) { entry in
            entry.state = .sending
            entry.transmissions += 1
            entry.lastSentAt = at
        }
    }

    func markFailed(platformMsgId: String) {
        mutate(platformMsgId) { entry in entry.state = .failed }
    }

    /// A rebased page hid the floor — this entry's absence is unknowable from the
    /// page alone. Park it `unknown` pending the point lookup.
    func markUnknown(platformMsgId: String) {
        mutate(platformMsgId) { entry in
            if entry.state != .failed { entry.state = .unknown }
        }
    }

    /// Point lookup proved the key absent: the normal retry machine resumes.
    /// `lastSentAt` is zeroed so the next sweep treats the entry as due
    /// immediately; the lookup itself consumes no transmission.
    func resumeSending(platformMsgId: String) {
        mutate(platformMsgId) { entry in
            entry.state = .sending
            entry.lastSentAt = 0
        }
    }

    /// Manual retry from the failed affordance: same key, the automatic-
    /// transmission budget AND the queued-lifetime clock reset (a hand retry is a
    /// fresh attempt — `createdAt` moves so the `queuedTimeout` cap starts over,
    /// not instantly re-fires off the original enrolment). Re-creates the entry
    /// if it was released.
    func resetForManualRetry(platformMsgId: String, text: String, attachments: [OutboxAttachment]) {
        let at = now().timeIntervalSince1970
        if var entry = items[platformMsgId] {
            entry.state = .sending
            entry.transmissions = 0
            entry.createdAt = at
            entry.lastSentAt = 0
            items[platformMsgId] = entry
        } else {
            items[platformMsgId] = OutboxEntry(
                platformMsgId: platformMsgId, text: text, attachments: attachments,
                state: .sending, transmissions: 0,
                createdAt: at, lastSentAt: 0)
        }
        persist()
    }

    private func mutate(_ platformMsgId: String, _ body: (inout OutboxEntry) -> Void) {
        guard var entry = items[platformMsgId] else { return }
        body(&entry)
        items[platformMsgId] = entry
        persist()
    }

    // MARK: - Persistence

    nonisolated static func directory(in supportDirectory: URL) -> URL {
        let dir = supportDirectory.appendingPathComponent("outbox", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private nonisolated static func fileURL(for sessionId: String, in directory: URL) -> URL {
        let sanitized = sessionId.replacingOccurrences(of: "/", with: "_")
        return directory.appendingPathComponent("\(sanitized).json")
    }

    private nonisolated static func read(
        sessionId: String, in directory: URL
    ) -> [String: OutboxEntry] {
        guard let data = try? Data(contentsOf: fileURL(for: sessionId, in: directory)),
            let list = try? JSONDecoder().decode([OutboxEntry].self, from: data)
        else { return [:] }
        return Dictionary(list.map { ($0.platformMsgId, $0) }, uniquingKeysWith: { _, b in b })
    }

    private func persist() {
        let url = Self.fileURL(for: sessionId, in: directory)
        if items.isEmpty {
            try? FileManager.default.removeItem(at: url)
            return
        }
        guard let data = try? JSONEncoder().encode(Array(items.values)) else { return }
        try? data.write(to: url, options: .atomic)
    }

    /// Wipe every session's outbox — called on logout/rebind alongside
    /// `TranscriptStore.deleteAll(in:)`.
    nonisolated static func deleteAll(in supportDirectory: URL) {
        try? FileManager.default.removeItem(at: directory(in: supportDirectory))
    }
}
